### PR TITLE
Jvenstad/deployment cleanup part deux

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden please review and discuss. 
@bratseth join in if you like.

I'm trying to make it easier to read and reason about this code, by reducing the required mental mapping (read: inlining stuff) and making logic more visible (read: renaming methods to reveal some of their logic, rather than just their result, which is often known by context). 

I hope you can point out where you disagree. 

I'll also point out the few places where the changes are not purely refactoring, but where I think the behavior should be unchanged.

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java [hunk 1]
```diff
@@ -126,42 +128,29 @@ public LockedApplication withOutstandingChange(boolean outstandingChange) {
...
 
-    private Version determineTriggerVersion(DeploymentJobs.JobType jobType, Controller controller) {
-        Optional<Zone> zone = jobType.zone(controller.system());
-        if ( ! zone.isPresent()) // a sloppy test TODO: Fix
-            return controller.systemVersion();
-        return currentDeployVersion(controller, zone.get());
+    private Version deployVersionFor(DeploymentJobs.JobType jobType, Controller controller) {
+        return jobType == JobType.component
+               ? controller.systemVersion()
+               : deployVersionIn(jobType.zone(controller.system()).get(), controller);
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java]: I believe these conditions are the same? The only JobType we ever trigger, which does not have a zone, is the component job, unless we are doing something very wrong.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java [hunk 2]
```diff
@@ -126,42 +128,29 @@ public LockedApplication withOutstandingChange(boolean outstandingChange) {
...
-    private Optional<ApplicationRevision> determineTriggerRevision(DeploymentJobs.JobType jobType,
-                                                                   Controller controller) {
-        Optional<Zone> zone = jobType.zone(controller.system());
-        if ( ! zone.isPresent()) // a sloppy test TODO: Fix
-            return Optional.empty();
-        return currentDeployRevision(jobType.zone(controller.system()).get());
+    private Optional<ApplicationRevision> deployRevisionFor(DeploymentJobs.JobType jobType, Controller controller) {
+        return jobType == JobType.component
+               ? Optional.empty()
+               : deployRevisionIn(jobType.zone(controller.system()).get());
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java]: Same as above.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 3]
```diff
@@ -254,39 +257,24 @@ public void cancelChange(ApplicationId applicationId) {
...
-        if (zone.isPresent() && jobType.isProduction()) {
-            // Skip triggering of jobs for zones where the application should not be deployed
-            if ( ! application.deploymentSpec().includes(jobType.environment(), Optional.of(zone.get().region()))) {
-                return false;
-            }
-        }
-        return true;
+    private boolean hasJob(JobType jobType, Application application) {
+        if ( ! jobType.isProduction()) return true; // Deployment spec only determines this for production jobs.
+        return application.deploymentSpec().includes(jobType.environment(), jobType.region(controller.system()));
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: `zone.isPresent()` seemed redundant, when `&&`-ed with `jobType.isProduction()`.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 4]
```diff
@@ -345,16 +333,16 @@ private boolean allowedTriggering(JobType jobType, LockedApplication application
         //       by instead basing the decision on what is currently deployed in the zone. However,
         //       this leads to some additional corner cases, and the possibility of blocking an application
         //       fix to a version upgrade, so not doing it now
-        if (jobType.isProduction() && application.deployingBlocked(clock.instant())) return false;
+        if (application.deploying().isPresent()) {
+            if (jobType.isProduction() && application.deploying().get().blockedBy(application.deploymentSpec(), clock.instant())) return false;
+            if (application.deploying().get() instanceof VersionChange &&
+                    isOnNewerVersionInProductionThan(((VersionChange) application.deploying().get()).version(), application, jobType)) return false;
+        }
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: Not sure this really helped readability >_<

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 5]
```diff
@@ -389,16 +377,16 @@ private boolean isProduction(JobType job) {
...
     private boolean acceptNewRevisionNow(LockedApplication application) {
         if ( ! application.deploying().isPresent()) return true;
+
         if ( application.deploying().get() instanceof Change.ApplicationChange) return true; // more changes are ok
 
         if ( application.deploymentJobs().hasFailures()) return true; // allow changes to fix upgrade problems
-        if ( application.isBlocked(clock.instant())) return true; // allow testing changes while upgrade blocked (debatable)
+
+        if ( ! application.deploymentSpec().canUpgradeAt(clock.instant()) ||
+             ! application.deploymentSpec().canChangeRevisionAt(clock.instant())) return true; // allow testing changes while upgrade blocked (debatable)
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: This seemed a bit strange, so inlined to make the strangness more apparent.

**Comment by bratseth** [comment]
Looks good to me .... 90% better, 10% worse  :-P

**Comment by jonmv** [comment]
Hehe, I'm not surprised. Would you care to point out where you think I f*ed up?

**Comment by bratseth** [comment]
Just the one where you point it out yourself.

**Comment by bratseth** [comment]
... and the inlining og application.isBlocked.

**Comment by mpolden** [comment]
Merge after the discussion converges.

**Comment by jonmv** [comment]
Yes, that last one was, as commented, to accentuate the apparent weirdness:

* We reach this condition only when we're currently upgrading:
    * We are upgrading; since we are currently blocking upgrades, we allow a new revision for system tests, and possibly further down.j
    * We are upgrading; since we are currently blocking revisions, we allow a new revision. *_Huh?_* Perhaps this makes sense, but how?

**Comment by bratseth** [comment]
The feature here is that if some developer is working during a block period, they will have their changes built and tested, even though an upgrade is in progress but blocked. It's debatable whether the advantage of this outweights the drawbacks, but I don't have a strong opinion and the stream team asked for it.

**Comment by jonmv** [comment]
But this doesn't check whether it's upgrades or revision changes that are blocked ... It's possible it all works out in the end, though?

**Comment by bratseth** [comment]
It should be the same in both cases. Blocking is just for production

**Comment by jonmv** [comment]
Right, so in both cases, the new revision may interfere with an ongoing upgrade _in testing_, but isn't let out into world. 

A minor gotcha, perhaps: the component job has to be triggered (manually) again, after the current upgrade finishes, since the new revision isn't stored as an outstanding change, and can't deploy to production unless an application change is `deploying()`.

**Comment by bratseth** [comment]
> A minor gotcha, perhaps: the component job has to be triggered (manually) again,

True atm maybe, but that will be fixed when we stop having a Change.ApplicationChange and rely on the revision state of jobs instead.

**Comment by jonmv** [comment]
It's not our problem, but perhaps something to remember to tell them, when they wonder why things aren't triggered :)

**Pull request merged**